apis/s3/src/main/java/org/jclouds/s3/filters/ChunkedUploadPayload.java

```diff
@@ -0,0 +1,213 @@
+        buffer.append(CHUNK_STRING_TO_SIGN_PREFIX);
+        buffer.append("\n");
+        buffer.append(timestamp).append("\n");
+        buffer.append(scope).append("\n");
+        buffer.append(lastComputedSignature).append("\n");
+        buffer.append(hex(hash(nonsigExtension))).append("\n");
+        buffer.append(hex(hash(dataToChunk)));
+
+        String chunkStringToSign = buffer.toString();
+
+        // compute the V4 signature for the chunk
+        String chunkSignature;
+        try {
+            chunkSignature = hex(readBytes(toInputStream(chunkStringToSign), hmacSHA256));
+        } catch (IOException e) {
+            throw new HttpException("hmac sha256 chunked signature error");
+        }
+
+        // cache the signature to include with the next chunk's signature computation
+        lastComputedSignature = chunkSignature;
+
+        // construct the actual chunk, comprised of the non-signed extensions, the
+        // 'headers' we just signed and their signature, plus a newline then copy
+        // that plus the user's data to a payload to be written to the request stream
+        chunkHeader.append(nonsigExtension + CHUNK_SIGNATURE_HEADER + chunkSignature);
+        chunkHeader.append(CLRF);
+
+        byte[] header = chunkHeader.toString().getBytes(UTF_8);
+        byte[] signedChunk = new byte[header.length + dataToChunk.length + TRAILER.length];
+        System.arraycopy(header, 0, signedChunk, 0, header.length);
+        // chunk-data
+        System.arraycopy(dataToChunk, 0, signedChunk, header.length, dataToChunk.length);
+        System.arraycopy(TRAILER, 0, signedChunk, header.length + dataToChunk.length, TRAILER.length);
+
+        // this is the total data for the chunk that will be sent to the request stream
+        return signedChunk;
+    }
+
+    @Override
+    public void release() {
+        this.payload.release();
+    }
+
+    @Override
+    public boolean isRepeatable() {
+        return this.payload.isRepeatable();
+    }
+
+    @Override
+    public InputStream openStream() throws IOException {
+        return new SequenceInputStream(new ChunkedInputStreamEnumeration(this.payload.openStream(), chunkedBlockSize));
+    }
+
+    private class ChunkedInputStreamEnumeration implements Enumeration<InputStream> {
+        private final InputStream inputStream;
+        private boolean lastChunked;
+        private byte[] buffer;
+
+        ChunkedInputStreamEnumeration(InputStream inputStream, int chunkedBlockSize) {
+            this.inputStream = new BufferedInputStream(inputStream, chunkedBlockSize);
+            buffer = new byte[chunkedBlockSize];
+            lastChunked = false;
+        }
+
+        @Override
+        public boolean hasMoreElements() {
+            return !lastChunked;
+        }
+
+        @Override
+        public InputStream nextElement() {
+            int bytesRead;
+            try {
+                bytesRead = inputStream.read(buffer, 0, buffer.length);
```

Comment:
bytesRead can be less than buffer.length!

This caused problems during our tests, when the underlying inputStream only allows 8k blocks. The result was, that the pre-calculated content-length (using 64k chunks) was less than the actual send content length (using 8k chunks got from the InputStream).

That leads to HttpUrlConnection's "too many bytes written" Exception later on.

We fixed that on our side with a wrapping InputStream that always returns the requested length when read(buf,off,len) is invoked. Performing n additional reads from the underlying inputStream if necessary.

This way, the expected 64k blocks were send and the pre-calculated content-length matched the actual content length.

Maybe a loop must be added here to read exactly buffer.length bytes from the inputStream matching the chunkedBlockSize, like we did in our inputStream wrapper.